typescript-requiring-type-checking.js

```diff
@@ -15,23 +15,23 @@ module.exports = {
         /*
             TypeScript rules outside of the recommended configuration
         */
-        // '@typescript-eslint/naming-convention': 'error',
-        // '@typescript-eslint/no-base-to-string': 'error',
-        // '@typescript-eslint/no-confusing-void-expression': 'error',
-        // '@typescript-eslint/no-unnecessary-boolean-literal-compare': 'error',
-        // '@typescript-eslint/no-unnecessary-condition': 'error',
-        // '@typescript-eslint/no-unnecessary-qualifier': 'error',
-        // '@typescript-eslint/no-unnecessary-type-arguments': 'error',
-        // '@typescript-eslint/non-nullable-type-assertion-style': 'error',
-        // '@typescript-eslint/prefer-includes': 'error',
-        // '@typescript-eslint/prefer-nullish-coalescing': 'error',
-        // '@typescript-eslint/prefer-readonly': 'error',
-        // '@typescript-eslint/prefer-readonly-parameter-types': 'error',
-        // '@typescript-eslint/prefer-reduce-type-parameter': 'error',
-        // '@typescript-eslint/prefer-string-starts-ends-with': 'error',
-        // '@typescript-eslint/promise-function-async': 'error',
-        // '@typescript-eslint/require-array-sort-compare': 'error',
-        // '@typescript-eslint/strict-boolean-expressions': 'error',
-        // '@typescript-eslint/switch-exhaustiveness-check': 'error'
+        '@typescript-eslint/naming-convention': 'error',
+        '@typescript-eslint/no-base-to-string': 'error',
+        '@typescript-eslint/no-confusing-void-expression': 'error',
+        '@typescript-eslint/no-unnecessary-boolean-literal-compare': 'error',
+        '@typescript-eslint/no-unnecessary-condition': 'error',
+        '@typescript-eslint/no-unnecessary-qualifier': 'error',
+        '@typescript-eslint/no-unnecessary-type-arguments': 'error',
+        '@typescript-eslint/non-nullable-type-assertion-style': 'error',
+        '@typescript-eslint/prefer-includes': 'error',
+        '@typescript-eslint/prefer-nullish-coalescing': 'error',
```

Comment:
[@typescript-eslint/prefer-nullish-coalescing](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/prefer-nullish-coalescing.md)
Actually, the examples for this one would be bugs. I don't use the nullish operator specifically to check for empty string. My argument against this one is similar to strict-boolean-expressions below. Non-boolean expressions are useful/awesome and very commonly used.

Reply:
@jattasNI to check if this rule conflicts with the rule that caused [Nicolas' PR](https://dev.azure.com/ni/DevCentral/_git/op-web-portal/pullrequest/155326) to remove instances of logical chaining.

In our in person discussion we were split between preferring || and ??, especially whether '' would typically be treated as invalid (in the same category as null or undefined) or valid (in the same category as non-empty string).

Reply:
The rule in Nicolas' PR was **[no-unused-expressions](https://eslint.org/docs/rules/no-unused-expressions)** which reports errors for code like this: `this.selectionChangedSubscription && this.selectionChangedSubscription.unsubscribe();`. The recommendation in that case is to use optional chaining (`.?`): `this.selectionChangedSubscription?.unsubscribe();`

`prefer-nullish-coalescing` only fires for `||` whereas the pattern in `no-unused-expressions` was `&&` so I don't think they're in conflict.

I'll put up proposals for the two options under discussion: please vote to see where we all stand.

Reply:
**Proposal** 
```
        /* When coalescing (providing a valid value to replace a possibly invalid value), 
         * prefer nullish coalescing (`??`) over logical OR (`||`)
         * as the former will only coalesce for nullish values (null or undefined) 
         * while the latter will coalesce for all falsy values (empty string, 0) which is more likely to be unintended.
         */
        '@typescript-eslint/prefer-nullish-coalescing': 'error',

Reply:
Tried to think of what one-liner patterns we would use to implement nullishOrEmpty if this rule is enabled and could only think of the following:
```ts
const myFunc = (input: string | null): void => {
    const coerceNull = input ?? 'default';

    // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
    const coerceNullOrEmpty1 = input || 'default';

    const coerceNullOrEmpty2 = (input === '' ? undefined : input) ?? 'default';

    const coerceNullOrEmpty3 = input === null || input === undefined || input === '' ? 'default' : input;

    const coerceNullOrEmpty4 = typeof input === 'string' && input !== '' ? input : 'default';
};
```
If we were not in a TS world then I'd lean to option 4 since it is very explicit about what is allowed; option 4 has the strongest guarantee that the result is a string.

Since we are in a TS world then option 4 is not necessary and the clearest representation of "nullishOrEmpty" may be the logical OR. Since we are in a TS world and assuming codebases do not proliferate `any` types, etc. then I guess it makes sense to allow logical OR coalescing...

Reply:
I voted on Jesse's but have an alternate proposal comment that is stricter and aligns with Trevor's strictNullChecks recommendation:
```
/*
    Both nullish coalescing (`??`) and logical OR (`||`) may be used 
    to provide default values in specific cases.

    Developers should choose the strictest operator for their use case:
    - `??` to provide a default for nullish values (null or undefined) 
    - `||` to provide a default for values that are nullish or one of: empty string / boolean false.

    Logical OR checks should not be used for other falsy values such as numbers (0, -0, 0n, NaN).
 */
'@typescript-eslint/prefer-nullish-coalescing': 'off',
```

Reply:
@rajsite this thread is ready to close using your revised comment above. @LukasKall voted opposite to where we landed in person. If Milan's argument isn't convincing enough for him to change his vote, we can revisit.

Reply:
I'm fine with Milan's proposal so this thread can be closed.